File: frontend/src/lib/sales.ts
import { apiClient } from './axios';

// ==========================================
// TYPES
// ==========================================

export enum SaleType {
  INVOICE = 'invoice',
  CREDIT_NOTE = 'credit_note',
}

export enum PaymentStatus {
  UNPAID = 'unpaid',
  PARTIAL = 'partial',
  PAID = 'paid',
}

export enum PaymentMethod {
  CASH = 'cash',
  MPESA = 'mpesa',
  BANK = 'bank',
  CARD = 'card',
}

export interface SaleItem {
  id: string;
  sale_id: string;
  product_variant_id: string;
  quantity: number;
  unit_price: number;
  discount_percentage: number;
  discount_amount: number;
  line_total: number;
  created_at: string;
}

export interface SaleItemWithDetails extends SaleItem {
  product_variant?: {
    id: string;
    variant_name: string;
    sku: string | null;
    products: {
      name: string;
    };
  };
}

export interface Sale {
  id: string;
  company_id: string;
  customer_id: string;
  sale_number: string;
  sale_type: SaleType;
  original_sale_id: string | null;
  sale_date: string;
  storage_location_id: string;
  subtotal: number;
  discount_percentage: number;
  discount_amount: number;
  total_amount: number;
  payment_status: PaymentStatus;
  amount_paid: number;
  amount_due: number;
  notes: string | null;
  created_by: string | null;
  created_at: string;
  updated_at: string;
}

export interface SaleWithDetails extends Sale {
  customer?: {
    id: string;
    name: string;
    customer_type: string;
    email?: string;
    phone?: string;
  };
  storage_location?: {
    id: string;
    name: string;
  };
  original_sale?: {
    sale_number: string;
    sale_type: string;
  };
  items: SaleItemWithDetails[];
  payments: SalePayment[];
}

export interface SalePayment {
  id: string;
  sale_id: string;
  payment_date: string;
  amount: number;
  payment_method: PaymentMethod;
  reference_number: string | null;
  notes: string | null;
  created_by: string | null;
  created_at: string;
}

export interface SaleItemCreateRequest {
  product_variant_id: string;
  quantity: number;
  unit_price: number;
  discount_percentage?: number;
}

export interface SaleCreateRequest {
  customer_id: string;
  sale_date: string;
  storage_location_id: string;
  items: SaleItemCreateRequest[];
  notes?: string;
}

export interface CreditNoteItemRequest {
  sale_item_id: string;
  return_quantity: number;
}

export interface CreditNoteCreateRequest {
  original_sale_id: string;
  sale_date: string;
  items: CreditNoteItemRequest[];
  notes?: string;
}

export interface SalePaymentCreateRequest {
  sale_id: string;
  payment_date: string;
  amount: number;
  payment_method: PaymentMethod;
  reference_number?: string;
  notes?: string;
}

// ==========================================
// API FUNCTIONS
// ==========================================

export const salesAPI = {
  // Get all sales
  getAll: async (params?: {
    sale_type?: SaleType;
    payment_status?: PaymentStatus;
    customer_id?: string;
    from_date?: string;
    to_date?: string;
    limit?: number;
  }): Promise<SaleWithDetails[]> => {
    const response = await apiClient.get<SaleWithDetails[]>('/sales/', { params });
    return response.data;
  },

  // Get single sale
  getById: async (id: string): Promise<SaleWithDetails> => {
    const response = await apiClient.get<SaleWithDetails>(`/sales/${id}`);
    return response.data;
  },

  // Create sale (invoice)
  create: async (data: SaleCreateRequest): Promise<Sale> => {
    const response = await apiClient.post<Sale>('/sales', data);
    return response.data;
  },

  // Create credit note
  createCreditNote: async (data: CreditNoteCreateRequest): Promise<Sale> => {
    const response = await apiClient.post<Sale>('/sales/credit-notes/', data);
    return response.data;
  },

  // Record payment
  recordPayment: async (data: SalePaymentCreateRequest): Promise<{
    message: string;
    payment: SalePayment;
    updated_sale: {
      payment_status: PaymentStatus;
      amount_paid: number;
      amount_due: number;
    };
  }> => {
    const response = await apiClient.post('/sales/payments', data);
    return response.data;
  },

  // Get payments for a sale
  getPayments: async (saleId: string): Promise<SalePayment[]> => {
    const response = await apiClient.get<SalePayment[]>(`/sales/payments/${saleId}`);
    return response.data;
  },

  // Download PDF
  downloadPDF: async (saleId: string, saleNumber: string): Promise<void> => {
    const response = await apiClient.get(`/sales/${saleId}/pdf/`, {
      responseType: 'blob',
    });
    
    // Create download link
    const url = window.URL.createObjectURL(new Blob([response.data]));
    const link = document.createElement('a');
    link.href = url;
    link.setAttribute('download', `${saleNumber}.pdf`);
    document.body.appendChild(link);
    link.click();
    link.remove();
    window.URL.revokeObjectURL(url);
  },
};

File: backend/app/schemas/sales.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime, date
from decimal import Decimal
from enum import Enum

# ==========================================
# ENUMS
# ==========================================

class SaleType(str, Enum):
    invoice = "invoice"
    credit_note = "credit_note"

class PaymentStatus(str, Enum):
    unpaid = "unpaid"
    partial = "partial"
    paid = "paid"

class PaymentMethod(str, Enum):
    cash = "cash"
    mpesa = "mpesa"
    bank = "bank"
    card = "card"

# ==========================================
# SALE SCHEMAS
# ==========================================

class SaleItemBase(BaseModel):
    product_variant_id: str
    quantity: int = Field(..., description="Positive for invoices, negative for credit notes")
    unit_price: Decimal = Field(..., ge=0)
    discount_percentage: Decimal = Field(default=0, ge=0, le=100)

class SaleItemCreate(SaleItemBase):
    pass

class SaleItemResponse(SaleItemBase):
    id: str
    sale_id: str
    discount_amount: Decimal
    line_total: Decimal
    created_at: datetime

    class Config:
        from_attributes = True

class SaleItemWithDetails(SaleItemResponse):
    product_variant: Optional[dict] = None

class SaleBase(BaseModel):
    customer_id: str
    sale_date: date = Field(default_factory=date.today)
    storage_location_id: str
    notes: Optional[str] = None

class SaleCreate(SaleBase):
    sale_type: SaleType = SaleType.invoice
    original_sale_id: Optional[str] = None
    items: List[SaleItemCreate] = Field(..., min_length=1)

class SaleResponse(SaleBase):
    id: str
    company_id: str
    sale_number: str
    sale_type: SaleType
    original_sale_id: Optional[str] = None
    subtotal: Decimal
    discount_percentage: Decimal
    discount_amount: Decimal
    total_amount: Decimal
    payment_status: PaymentStatus
    amount_paid: Decimal
    amount_due: Decimal
    created_by: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

class SaleWithDetails(SaleResponse):
    customer: Optional[dict] = None
    storage_location: Optional[dict] = None
    original_sale: Optional[dict] = None
    items: List[SaleItemWithDetails] = []
    payments: List[dict] = []

# ==========================================
# PAYMENT SCHEMAS
# ==========================================

class SalePaymentBase(BaseModel):
    sale_id: str
    payment_date: date = Field(default_factory=date.today)
    amount: Decimal = Field(..., gt=0)
    payment_method: PaymentMethod
    reference_number: Optional[str] = None
    notes: Optional[str] = None

class SalePaymentCreate(SalePaymentBase):
    pass

class SalePaymentResponse(SalePaymentBase):
    id: str
    created_by: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True

# ==========================================
# CREDIT NOTE SCHEMAS
# ==========================================

class CreditNoteItemBase(BaseModel):
    sale_item_id: str
    return_quantity: int = Field(..., gt=0)

class CreditNoteCreate(BaseModel):
    original_sale_id: str
    sale_date: date = Field(default_factory=date.today)
    items: List[CreditNoteItemBase] = Field(..., min_length=1)
    notes: Optional[str] = None